meilisearch-core/src/facets.rs

```diff
@@ -0,0 +1,342 @@
+    pub fn new(field_id: FieldId, value: String) -> Self {
+        let value = match value.cow_to_lowercase() {
+                Cow::Borrowed(_) => value,
+                Cow::Owned(s) => s,
+        };
+        Self(field_id, value)
+    }
+
+    fn from_str(
+        s: &str,
+        schema: &Schema,
+        attributes_for_faceting: &[FieldId],
+    ) -> Result<Self, FacetError> {
+        let mut split = s.splitn(2, ":");
+        let key = split
+            .next()
+            .ok_or_else(|| FacetError::InvalidFormat(s.to_string()))?;
+        let field_id = schema
+            .id(key)
+            .ok_or_else(|| FacetError::AttributeNotFound(key.to_string()))?;
+
+        if !attributes_for_faceting.contains(&field_id) {
+            return Err(FacetError::AttributeNotSet {
+                found: key.to_string(),
+                expected:  attributes_for_faceting
+                    .iter()
+                    .filter_map(|&id| schema.name(id))
+                    .map(str::to_string)
+                    .collect::<Vec<_>>() })
+        }
+        let value = split
+            .next()
+            .ok_or_else(|| FacetError::InvalidFormat(s.to_string()))?;
+        // unquoting the string if need be:
+        let mut indices = value.char_indices();
+        let value =  match (indices.next(), indices.last()) {
+            (Some((s, '\'')), Some((e, '\''))) |
+            (Some((s, '\"')), Some((e, '\"'))) => value[s + 1..e].to_string(),
+            _ => value.to_string(),
+        };
+        Ok(Self::new(field_id, value))
+    }
+}
+
+impl<'a> heed::BytesEncode<'a> for FacetKey {
+    type EItem = FacetKey;
+
+    fn bytes_encode(item: &'a Self::EItem) -> Option<Cow<'a, [u8]>> {
+        let mut buffer = Vec::with_capacity(2 + item.1.len());
+        let field_id_bytes: [u8; 2] = u16::from(item.0).to_be_bytes();
+        buffer.extend_from_slice(&field_id_bytes);
+        buffer.extend_from_slice(item.1.as_bytes());
+        Some(Cow::Owned(buffer))
+    }
+}
+
+impl<'a> heed::BytesDecode<'a> for FacetKey {
+    type DItem = FacetKey;
+
+    fn bytes_decode(bytes: &'a [u8]) -> Option<Self::DItem> {
+        let field_id_u16 = u16::from_be_bytes([bytes[0], bytes[1]]);
+        let field_id = FieldId(field_id_u16);
+        let string = Str::bytes_decode(&bytes[2..])?;
+        Some(FacetKey(field_id, string.to_string()))
+    }
+}
+
+pub fn add_to_facet_map(
+    facet_map: &mut HashMap<FacetKey, Vec<DocumentId>>,
+    field_id: FieldId,
+    value: Value,
+    document_id: DocumentId,
+) -> Result<(), FacetError> {
+    let value = match value {
+        Value::String(s) => s,
+        Value::Object(_) | Value::Array(_) => return Err(FacetError::InvalidDocumentAttribute(document_id)),
+        v => v.to_string(),
+    };
+    let key = FacetKey::new(field_id, value);
+    facet_map.entry(key).or_insert(Vec::new()).push(document_id);
+    Ok(())
+}
+
+pub fn facet_map_from_docids<'txn>(
+    rtxn: &'txn heed::RoTxn<MainT>,
```

Comment:
When you don't bound any returned type to the lifetime of the transaction you can avoid giving it a name.
```suggestion
pub fn facet_map_from_docids(
    rtxn: &heed::RoTxn<MainT>,
```